### PR TITLE
feat: implement Web Cache API

### PR DESCRIPTION
**Pull request by satyarohith**
This PR implements the Cache API for Deno.

The following APIs are implemented:

- `CacheStorage::open()`
- `CacheStorage::has()`
- `CacheStorage::delete()`
- `Cache::match()`
- `Cache::put()`
- `Cache::delete()`

Cache APIs don't support the [query options](https://w3c.github.io/ServiceWorker/#dictdef-cachequeryoptions) yet.

Spec: https://w3c.github.io/ServiceWorker/#cache-interface

### CODE REVIEW

**Review by littledivy — COMMENTED** [review]
I know this isn't ready but please also add a benchmark in `cli/bench/` like https://github.com/denoland/deno/blob/main/cli/bench/op_now.js

File: cli/worker.rs [hunk 1]
```diff
@@ -381,13 +381,18 @@ pub async fn create_main_worker(
...
+  let origin_storage_dir = maybe_storage_key.as_ref().map(|key| {
     ps.dir
       .root
       // TODO(@crowlKats): change to origin_data for 2.0
       .join("location_data")
       .join(checksum::gen(&[key.as_bytes()]))
   });
+  let cache_storage_dir = maybe_storage_key.map(|key| {
+    std::env::temp_dir()
+      .join("deno_cache")
```
Comment (bartlomieju) [inline cli/worker.rs:393]: Why are you using a temporary directory? It will go away after the program finishes, isn't this cache supposed to be persistent?

File: core/resources.rs [hunk 2]
```diff
@@ -134,6 +134,10 @@ impl ResourceTable {
   /// Returns a unique resource ID, which acts as a key for this resource.
   pub fn add_rc<T: Resource>(&mut self, resource: Rc<T>) -> ResourceId {
     let resource = resource as Rc<dyn Resource>;
+    self.add_rc_dyn(resource)
+  }
+
+  pub fn add_rc_dyn(&mut self, resource: Rc<dyn Resource>) -> ResourceId {
```
Comment (bartlomieju) [inline core/resources.rs:140]: Why is this API needed?

File: ext/cache/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# deno_cache
+
+This crate implements the Cache API.
+
+Spec: https://console.spec.whatwg.org/
```
Comment (bartlomieju) [inline ext/cache/README.md:5]: Wrong link

File: runtime/web_worker.rs [hunk 4]
```diff
@@ -392,6 +393,7 @@ impl WebWorker {
         file_fetch_handler: Rc::new(deno_fetch::FsFetchHandler),
         ..Default::default()
       }),
+      deno_cache::init::<SqliteBackedCache>(None).disable(),
```
Comment (bartlomieju) [inline runtime/web_worker.rs:396]: Why add it if it's always disabled?

File: ext/cache/01_cache.js [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  const { getHeader } = window.__bootstrap.headers;
+
+  class CacheStorage {
+    constructor() {
+      return this;
+    }
+
+    async open(cacheName) {
+      const cacheId = await core.opAsync("op_cache_storage_open", cacheName);
+      return PromiseResolve(new Cache(cacheId));
```
Comment (Phosra) [inline ext/cache/01_cache.js:25]: PromiseResolve/PromiseReject are unnecessary in async functions.

File: ext/cache/01_cache.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+      this.#id = cacheId;
+      return this;
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#dom-cache-put */
+    async put(request, response) {
+      // Step 1.
+      let innerRequest = null;
+      // Step 2.
+      if (request instanceof Request) {
```
Comment (Phosra) [inline ext/cache/01_cache.js:50]: iirc, the usual idiom is
```suggestion
      if (ObjectPrototypeIsPrototypeOf(RequestPrototype, request) {
```

File: ext/cache/01_cache.js [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        innerRequest = toInnerRequest(request);
+      } else {
+        // Step 3.
+        try {
+          innerRequest = toInnerRequest(new Request(request));
+        } catch (error) {
+          return PromiseReject(error);
```
Comment (Phosra) [inline ext/cache/01_cache.js:57]: Ditto; just remove the try/catch?

File: ext/cache/01_cache.js [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        }
+      }
+      // Step 4.
+      const reqUrl = innerRequest.url();
+      if (!reqUrl.startsWith("http:") && !reqUrl.startsWith("https:")) {
+        const url = new URL(reqUrl);
+        if (url.protocol !== "http:" && url.protocol !== "https:") {
+          return PromiseReject(
```
Comment (Phosra) [inline ext/cache/01_cache.js:65]: Ditto; throw?

File: ext/cache/01_cache.js [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+          innerRequest = toInnerRequest(new Request(request));
+        } catch (error) {
+          return PromiseReject(error);
+        }
+      }
+      // Step 4.
+      const reqUrl = innerRequest.url();
+      if (!reqUrl.startsWith("http:") && !reqUrl.startsWith("https:")) {
+        const url = new URL(reqUrl);
+        if (url.protocol !== "http:" && url.protocol !== "https:") {
```
Comment (Phosra) [inline ext/cache/01_cache.js:64]: When wouldn't it start with these, yet URL would consider these the protocol?

File: ext/cache/01_cache.js [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+          requestUrl: innerRequest.url(),
+          responseHeaders: innerResponse.headerList,
+          requestHeaders: innerRequest.headerList,
+          responseHasBody: innerResponse.body !== null,
+          responseStatus: innerResponse.status,
+          responseStatusText: innerResponse.statusMessage,
+        },
+      );
+      if (innerResponse.body) {
+        const reader = innerResponse.body.stream.getReader();
```
Comment (Phosra) [inline ext/cache/01_cache.js:117]: Not familiar with this section and haven't looked around too well at the moment (sorry), does the reader evoke user code?

If it's a possibility that this is a user-provided `ReadableStream`, then I'll double-check and cross-reference this spot with the spec.

File: ext/cache/01_cache.js [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+                  // error.
+                  controller.error(err);
+                  controller.close();
+                  core.close(responseBodyRid);
+                }
+              },
+            });
+          }
+
+          const response = new Response(
```
Comment (Phosra) [inline ext/cache/01_cache.js:240]: I think there are internal fastpaths for constructing `Response`/`Request`?

I'll check later.

**Comment by Phosra** [comment]
Unless I'm mistaken, why is `matchAll` not going to be implemented?

File: cli/worker.rs (hunk 1 above)
Comment (satyarohith) [reply cli/worker.rs:393]: I didn't want to do quota management for the first iteration so I opted for tmp directory

File: core/resources.rs (hunk 2 above)
Comment (satyarohith) [reply core/resources.rs:140]: Structs (e.g. `SqliteBackedCache`) that implement the `Cache` trait return dynamic resources, so I wanted this API to add them to the state.

File: runtime/web_worker.rs (hunk 4 above)
Comment (satyarohith) [reply runtime/web_worker.rs:396]: Cache API is supported in WebWorker context, I've enabled it.

File: ext/cache/lib.rs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct CacheDeleteRequest {
+  pub cache_id: i64,
+  pub request_url: String,
+}
+
+#[op]
+pub async fn op_cache_put<CA>(
+  state: Rc<RefCell<OpState>>,
+  request_response: CachePutRequest,
```
[the remark this answers is not in the document]
Comment (satyarohith) [reply ext/cache/lib.rs:121]: I would like to skip performance optimizations to the next iteration. :)

**Review by satyarohith — COMMENTED** [review]
Most WPT tests for the API are now passing. The failing tests are mostly due to missing support for cache query options, or tests that depend on origin, or tests that depend on missing APIs like cache.add().

```
➜  deno git:(ext/cache) ./tools/wpt.ts run -- service-workers/cache-storage/cache-put.https.any.html
    Finished dev [unoptimized + debuginfo] target(s) in 0.27s
Going to run 1 test files.
----------------------------------------
/service-workers/cache-storage/cache-put.https.any.html

test Cache.put called with simple Request and Response ... ok
test Cache.put called with Request and Response from fetch() ... failed (expected)
test Cache.put with Request without a body ... ok
test Cache.put with Response without a body ... ok
test Cache.put with a Response containing an empty URL ... ok
test Cache.put with an empty response body ... ok
test Cache.put with synthetic 206 response ... ok
test Cache.put with HTTP 206 response ... ok
test Cache.put with opaque-filtered HTTP 206 response ... failed (expected)
test Cache.put with HTTP 500 response ... failed (expected)
test Cache.put called twice with matching Requests and different Responses ... ok
test Cache.put called twice with request URLs that differ only by a fragment ... ok
test Cache.put with a string request ... ok
test Cache.put with an invalid response ... ok
test Cache.put with a non-HTTP/HTTPS request ... ok
test Cache.put with a relative URL ... ok
test Cache.put with a non-GET request ... ok
test Cache.put with a null response ... ok
test Cache.put with a POST request ... ok
test Cache.put with a used response body ... ok
test getReader() after Cache.put ... ok
test Cache.put with a VARY:* Response ... ok
test Cache.put with an embedded VARY:* Response ... ok
test Cache.put with a VARY:* opaque response should not reject ... failed (expected)
test Cache.put should store Response.redirect() correctly ... ok
test Cache.put called with simple Request and blob Response ... ok
test Cache.put called with simple Request and form data Response ... ok


file result: ok. 23 passed; 0 failed; 4 expected failure; total 27

========================================

final result: ok. 1 passed; 0 failed; 0 expected failure; total 1

➜  deno git:(ext/cache) ./tools/wpt.ts run -- service-workers/cache-storage/cache-match.https.any.html
    Finished dev [unoptimized + debuginfo] target(s) in 0.29s
Going to run 1 test files.
----------------------------------------
/service-workers/cache-storage/cache-match.https.any.html

test Cache.match with no matching entries ... ok
test Cache.match with URL ... ok
test Cache.match with Request ... ok
test Cache.match with multiple cache hits ... ok
test Cache.match with new Request ... ok
test Cache.match with HEAD ... ok
test Cache.match with ignoreSearch option (request with no search parameters) ... ok
test Cache.match with ignoreSearch option (request with search parameter) ... ok
test Cache.match supports ignoreMethod ... failed (expected)
test Cache.match supports ignoreVary ... failed (expected)
test Cache.match does not support cacheName option ... ok
test Cache.match with URL containing fragment ... ok
test Cache.match with string fragment "http" as query ... ok
test Cache.match with responses containing "Vary" header ... ok
test Cache.match with Request and Response objects with different URLs ... failed (expected)
test Cache.match invoked multiple times for the same Request/Response ... ok
test Cache.match blob should be sliceable ... ok
test Cache.match with POST Request ... ok
test Cache.match with a non-2xx Response ... ok
test Cache.match with a network error Response ... failed (expected)
test Cache produces large Responses that can be cloned and read correctly. ... ok
test cors-exposed header should be stored correctly. ... failed (expected)
test MIME type should be set from content-header correctly. ... failed (expected)
test MIME type should reflect Content-Type headers of response. ... ok
test Cache.match ignores vary headers on opaque response. ... failed (expected)


file result: ok. 18 passed; 0 failed; 7 expected failure; total 25

========================================

final result: ok. 1 passed; 0 failed; 0 expected failure; total 1

➜  deno git:(ext/cache) ./tools/wpt.ts run -- service-workers/cache-storage/cache-delete.https.any.html
    Finished dev [unoptimized + debuginfo] target(s) in 0.29s
Going to run 1 test files.
----------------------------------------
/service-workers/cache-storage/cache-delete.https.any.html

test Cache.delete with no arguments ... ok
test Cache.delete called with a string URL ... ok
test Cache.delete called with a Request object ... ok
test Cache.delete called with a HEAD request ... failed (expected)
test Cache.delete supports ignoreVary ... failed (expected)
test Cache.delete with a non-existent entry ... ok
test Cache.delete with ignoreSearch option (request with search parameters) ... failed (expected)
test Cache.delete with ignoreSearch option (when it is specified as false) ... failed (expected)


file result: ok. 4 passed; 0 failed; 4 expected failure; total 8

========================================

```

File: ext/cache/01_cache.js [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+  class CacheStorage {
+    constructor() {
+      webidl.illegalConstructor();
+    }
+
+    async open(cacheName) {
+      webidl.assertBranded(this, CacheStoragePrototype);
+      const prefix = "Failed to execute 'open' on 'CacheStorage'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:30]: normalize `cacheName` with the DOMString webidl converter

File: ext/cache/01_cache.js [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+      const cacheId = await core.opAsync("op_cache_storage_open", cacheName);
+      return new Cache(cacheId);
+    }
+
+    async has(cacheName) {
+      webidl.assertBranded(this, CacheStoragePrototype);
+      const prefix = "Failed to execute 'has' on 'CacheStorage'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:38]: normalize `cacheName` with the DOMString webidl converter

File: ext/cache/01_cache.js [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+      return await core.opAsync("op_cache_storage_has", cacheName);
+    }
+
+    async delete(cacheName) {
+      webidl.assertBranded(this, CacheStoragePrototype);
+      const prefix = "Failed to execute 'delete' on 'CacheStorage'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:45]: normalize `cacheName` with the DOMString webidl converter

File: ext/cache/01_cache.js [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+    [_id];
+
+    constructor(cacheId) {
+      this[_id] = cacheId;
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#dom-cache-put */
+    async put(request, response) {
+      const prefix = "Failed to execute 'put' on 'Cache'";
+      webidl.requiredArguments(arguments.length, 2, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:63]: normalize the arguments with the appropriate webidl converters

File: ext/cache/01_cache.js [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+          }
+        }
+      }
+      // Step 12-19: TODO(@satyarohith): do the insertion in background.
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#cache-match */
+    async match(request, options) {
+      const prefix = "Failed to execute 'match' on 'Cache'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:142]: normalize the arguments with the appropriate webidl converters

File: ext/cache/01_cache.js [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+        return p[0];
+      } else {
+        return undefined;
+      }
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#cache-delete */
+    async delete(request, _options) {
+      const prefix = "Failed to execute 'delete' on 'Cache'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
```
Comment (crowlKats) [inline ext/cache/01_cache.js:154]: normalize the arguments with the appropriate webidl converters

**Review by crowlKats — DISMISSED** [review]

**Review by satyarohith — COMMENTED** [review]
Current benchmarks:
```
➜  deno git:(ext/cache) ./target/debug/deno bench --unstable -A ./cli/bench/cache_api.js
cpu: Apple M1 Pro
runtime: deno 1.25.4 (aarch64-apple-darwin)

file:///Users/sr/c/denoland/deno/cli/bench/cache_api.js
benchmark                  time (avg)             (min … max)       p75       p99      p995
------------------------------------------------------------- -----------------------------
cache_storage_open      61.95 µs/iter  (57.67 µs … 544.38 µs)  59.75 µs 107.96 µs 125.58 µs
cache_storage_has       59.27 µs/iter  (57.71 µs … 152.42 µs)  59.12 µs   64.5 µs  70.79 µs
cache_storage_delete     60.7 µs/iter  (57.88 µs … 228.04 µs)  59.71 µs  96.58 µs 101.08 µs
cache_put_body_10_KiB   24.96 ms/iter   (19.43 ms … 95.18 ms)  24.13 ms  95.18 ms  95.18 ms
cache_put_no_body      417.38 µs/iter   (382.29 µs … 2.35 ms)  411.5 µs 578.17 µs 654.33 µs
cache_match            176.12 µs/iter   (160.54 µs … 1.18 ms) 173.12 µs 266.67 µs 340.38 µs
cache_delete            80.26 µs/iter  (77.83 µs … 483.33 µs)  79.83 µs  90.62 µs  94.79 µs
```

**Comment by bartlomieju** [comment]
What's the reason for not adding these APIs?

- Cache::match()
- Cache::put()
- Cache::delete()

**Comment by satyarohith** [comment]
> What's the reason for not adding these APIs?
> 
> * Cache::match()
> * Cache::put()
> * Cache::delete()

They're supported. The structure of the readme might have confused you. I've updated it.

File: ext/cache/lib.rs [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+use std::path::PathBuf;
+use std::rc::Rc;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct CachePutRequest {
+  pub cache_id: i64,
+  pub request_url: String,
+  pub request_headers: Vec<(String, String)>,
+  pub response_headers: Vec<(String, String)>,
```
Comment (lucacasonato) [inline ext/cache/lib.rs:28]: These should both be `ByteString`. See #10991

File: ext/cache/sqlite.rs [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+        let db = db.lock();
+        db.query_row(
+          "INSERT OR REPLACE INTO request_response_list
+               (cache_id, request_url, request_headers, response_headers, response_body_key, response_status, response_status_text)
+               VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
+               RETURNING response_body_key",
+          (
+            request_response.cache_id,
+            &request_response.request_url,
+            serde_json::to_string(&request_response.request_headers)?,
```
Comment (lucacasonato) [inline ext/cache/sqlite.rs:172]: JSON serialization/deserialization is relatively slow. Also we can't use `ByteString` with `serde_json`. We'll need some other encoding for `Vec<(ByteString, ByteString)>` -> `Vec<u8>`.

File: ext/cache/sqlite.rs [hunk 21]
```diff
@@ -0,0 +1,401 @@
...
+          ),
+          |row| {
+            let response_body_key: Option<String> = row.get(0)?;
+            Ok(response_body_key)
+          },
+        )?
+      };
+      if let Some(body_key) = maybe_response_body {
+        let responses_dir  = cache_storage_dir.join(request_response.cache_id.to_string()).join("responses");
+          std::fs::create_dir_all(&responses_dir)?;
```
Comment (lucacasonato) [inline ext/cache/sqlite.rs:186]: Let's do this on cache open, rather than every insert.

File: ext/cache/sqlite.rs [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+        if cache_dir.exists() {
+          std::fs::remove_dir_all(cache_dir)?;
+        }
+      }
+      Ok::<bool, AnyError>(maybe_cache_id.is_some())
+    })
+    .await?
+  }
+
+  async fn put(
```
Comment (lucacasonato) [inline ext/cache/sqlite.rs:150]: This `put` is racy. If a user inserts an item into the cache with a pending body, then attempts to read that item from the cache, the item read will either error (no such file), or will return partial body data.

We need to change how puts work:
1. When calling `cache.put()` you create a resource on disk for that cache, with an arbitrary name.
2. Then create a body writer resource containing the file handle, request / resp metadata, and a timestamp indicating when the `cache.put` was started.
3. The body is then written to the file.
4. When the body stream is closed, you insert the metadata into sqlite, _if_ the timestamp in the resource is larger than the last inserted timestamp for that URL in sqlite.

This prevents the race, as other callers will not be able to see the entry until the entire body data is written. Additionally this prevents a scenario where two `cache.put` operations are started in close proximity, but the one that is later finishes before the earlier one, and then the data from the later operation is lost.

File: tools/wpt/expectation.json [hunk 23]
```diff
@@ -4732,5 +4732,278 @@
...
+        "Cache.match supports ignoreMethod",
+        "Cache.match supports ignoreVary",
+        "Cache.match with Request and Response objects with different URLs",
+        "Cache.match with a network error Response",
+        "cors-exposed header should be stored correctly.",
+        "MIME type should be set from content-header correctly.",
+        "Cache.match ignores vary headers on opaque response."
+      ],
+      "cache-delete.https.any.html": [
+        "Cache.delete called with a HEAD request",
```
Comment (lucacasonato) [inline tools/wpt/expectation.json:4947]: This sounds like it should be passing.

File: tools/wpt/expectation.json [hunk 24]
```diff
@@ -4732,5 +4732,278 @@
...
+        "CacheStorage.open with existing cache",
+        "CacheStorage names are DOMStrings not USVStrings"
+      ],
+      "common.https.window.html": true
+    },
+    "service-worker": {
+      "fetch-request-xhr-sync-error.https.window.html": false,
+      "ready.https.window.html": false,
+      "xhr-content-length.https.window.html": false
+    }
```
Comment (lucacasonato) [inline tools/wpt/expectation.json:5007]: remove

File: ext/cache/lib.rs [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+where
+  CA: Cache + 'static,
+{
+  let state = state.borrow();
+  let cache = state.try_borrow::<CA>().ok_or_else(|| {
+    custom_error(
+      "NotSupported",
+      "Cache API is not supported in this context.",
+    )
+  })?;
```
Comment (lucacasonato) [inline ext/cache/lib.rs:178]: I don't think this should be a `Deno.errors.NotSupported` error. Should be a `DOMExceptionNotSupportedError` instead.

File: cli/worker.rs (hunk 1 above)
Comment (lucacasonato) [reply cli/worker.rs:393]: Please add a TODO for quota management.

**Review by lucacasonato — COMMENTED** [review]
Looking good.

We currently the `SqliteBackedCache` is created unconditionally on startup. This involves creating an fs dir, and an sqlite db. Let's only create it on the first access of an op that requires the cache. This reduces startup cost, and defers issues for where the cache dir may not be writable to the actual call site of the cache API.

File: ext/cache/sqlite.rs (hunk 20 above)
Comment (satyarohith) [reply ext/cache/sqlite.rs:172]: What can be used here?

File: tools/wpt/expectation.json (hunk 23 above)
Comment (satyarohith) [reply tools/wpt/expectation.json:4947]: `ignoreMethod` option is not supported hence the test fails

File: ext/cache/sqlite.rs (hunk 20 above)
Comment (lucacasonato) [reply ext/cache/sqlite.rs:172]: We could do a simple custom serialisation in the form of `[ len name (u32) | name bytes | len val (u32) | val bytes ]*`. If there is something off the shelf we can use, I'd prefer that.

Comment (littledivy) [reply ext/cache/sqlite.rs:172]: Serialize into a buffer with V8 serializer? I'm not sure how stable that will be.

Comment (lucacasonato) [reply ext/cache/sqlite.rs:172]: That is stable.

File: ext/cache/sqlite.rs (hunk 22 above)
Comment (satyarohith) [reply ext/cache/sqlite.rs:153]: > When the body stream is closed, you insert the metadata into sqlite, if the timestamp in the resource is larger than the last inserted timestamp for that URL in sqlite.

What role does the timestamp play? I don't understand its role. Timestamp in `cache.put` will always be greater than the one in sqlite.

File: ext/cache/sqlite.rs (hunk 20 above)
Comment (satyarohith) [reply ext/cache/sqlite.rs]: Since this isn't a hard blocker, I would like to add a TODO and do the optimization later. Any objections?

File: ext/cache/sqlite.rs (hunk 22 above)
Comment (satyarohith) [reply ext/cache/sqlite.rs:156]: added a timestamp to track

File: ext/cache/sqlite.rs [hunk 26]
```diff
@@ -153,48 +159,30 @@ impl Cache for SqliteBackedCache {
...
-    }).await??;
-    if let Some(path) = maybe_body_path {
-      // TODO(@satyarohith): can we create the file in the blocking task above?
-      let file = tokio::fs::File::create(path).await?;
-      Ok(Some(Rc::new(CachePutResource::new(file))))
+    let response_body_key = if request_response.response_has_body {
+      Some(hash(&request_response.request_url))
+    } else {
+      None
+    };
```
Comment (lucacasonato) [inline ext/cache/sqlite.rs:166]: Currently the hash is just based off the request URL. This still results in a race:

1. `cache.put` with req + resp1 starts
2. `cache.put` with resp1 completes
3. `cache.match` with req. This returns resp1.
4. `cache.put` with req + resp2 starts
5. `cache.match` with req. This returns resp1, but with a corrupt body.

To fix this, we need to use a unique string as the key for the storage file for each request. Maybe just a v4 UUID.

**Review by lucacasonato — COMMENTED** [review]
Looking good. One last comment.

File: ext/cache/sqlite.rs (hunk 26 above)
Comment (satyarohith) [reply ext/cache/sqlite.rs:166]: Oh, I forgot to fix this. I'll use `format!("{request_url_hash}_{time_in_ns}")` as the key.

File: .gitmodules [hunk 27]
```diff
@@ -9,4 +9,3 @@
 [submodule "test_util/wpt"]
 	path = test_util/wpt
 	url = https://github.com/web-platform-tests/wpt.git
-
```
Comment (bartlomieju) [inline .gitmodules:12]: Nit: revert this change

File: cli/Cargo.toml [hunk 28]
```diff
@@ -27,6 +27,7 @@ path = "./bench/lsp_bench_standalone.rs"
 
 [build-dependencies]
 deno_broadcast_channel = { version = "0.64.0", path = "../ext/broadcast_channel" }
+deno_cache = { version = "0.1.0", path = "../ext/cache" }
```
Comment (bartlomieju) [inline cli/Cargo.toml:30]: @dsherret just to double check, do we need to publish manually first? Or are the automation scripts handling this?

File: cli/bench/cache_api.js [hunk 29]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+const cacheName = "cache-v1";
+const cache = await caches.open(cacheName);
+const req = "https://deno.com";
+
+Deno.bench(
+  `cache_storage_open`,
+  { n: 5e2 },
```
Comment (bartlomieju) [inline cli/bench/cache_api.js:9]: These do nothing (we run benches as long as there are statistically significant variations between runs). Suggest to remove all of these

File: cli/dts/lib.deno.window.d.ts [hunk 30]
```diff
@@ -83,6 +85,8 @@ declare var onunhandledrejection:
 declare var localStorage: Storage;
 /** @category Web Storage API */
 declare var sessionStorage: Storage;
+/** @category Cache API */
+declare var caches: CacheStorage;
```
Comment (bartlomieju) [inline cli/dts/lib.deno.window.d.ts:89]: Just to double check, it's on purpose that caches are only available in the main worker and not web workers?

File: ext/cache/README.md [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+
+- `CacheStorage::open()`
+- `CacheStorage::has()`
+- `CacheStorage::delete()`
+
+Cache APIs don't support the [query options][queryoptions] yet.
+
+- `Cache::match()`
+- `Cache::put()`
+- `Cache::delete()`
```
Comment (bartlomieju) [inline ext/cache/README.md:15]: Is this up to date?

File: ext/cache/lib.rs [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+use std::sync::Arc;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CachePutRequest {
+  pub cache_id: i64,
+  pub request_url: String,
+  // TODO(@satyarohith): use ByteString for headers.
+  // Note: ByteString cannot be stored as JSON in sqlite.
+  // Serialize into a buffer with V8 serializer.
```
Comment (bartlomieju) [inline ext/cache/lib.rs:29]: Just FYI serialization format is not stable across V8 releases

File: runtime/web_worker.rs [hunk 33]
```diff
@@ -373,6 +376,10 @@ impl WebWorker {
         Ok(())
       })
       .build();
+    let create_cache = options.cache_storage_dir.map(|storage_dir| {
+      let create_cache_fn = move || SqliteBackedCache::new(storage_dir.clone());
+      CreateCache(Arc::new(Box::new(create_cache_fn)))
```
Comment (bartlomieju) [inline runtime/web_worker.rs:381]: This seems off, if you use `Arc` you shouldn't need to box it first

File: runtime/worker.rs [hunk 34]
```diff
@@ -131,6 +134,10 @@ impl MainWorker {
       })
       .build();
     let exit_code = ExitCode(Arc::new(AtomicI32::new(0)));
+    let create_cache = options.cache_storage_dir.map(|storage_dir| {
+      let create_cache_fn = move || SqliteBackedCache::new(storage_dir.clone());
+      CreateCache(Arc::new(Box::new(create_cache_fn)))
```
Comment (bartlomieju) [inline runtime/worker.rs:139]: Ditto

**Review by bartlomieju — COMMENTED** [review]
Updating `test_util/std/` and `third_party/` submodules in this PR seems off, please revert these changes.

File: ext/cache/lib.rs (hunk 32 above)
Comment (lucacasonato) [reply ext/cache/lib.rs:29]: The format is not stable, but it is backwards compatible.

File: cli/dts/lib.deno.window.d.ts (hunk 30 above)
Comment (lucacasonato) [reply cli/dts/lib.deno.window.d.ts:89]: The API does exist for workers at runtime, so just a typings issue I think.

**Review by lucacasonato — COMMENTED** [review]
Are the `test_util/std` and `third_party` updates intentional?

File: .gitmodules [hunk 35]
```diff
@@ -9,4 +9,3 @@
 [submodule "test_util/wpt"]
 	path = test_util/wpt
 	url = https://github.com/web-platform-tests/wpt.git
```
Comment (satyarohith) [inline .gitmodules:11]:
```suggestion
	url = https://github.com/web-platform-tests/wpt.git

```

**Comment by satyarohith** [comment]
> Are the `test_util/std` and `third_party` updates intentional?

Nope. I reverted the changes.

File: runtime/web_worker.rs (hunk 33 above)
Comment (satyarohith) [reply runtime/web_worker.rs]: Updated - thanks!

File: cli/dts/lib.deno.window.d.ts (hunk 30 above)
Comment (satyarohith) [reply cli/dts/lib.deno.window.d.ts:89]: updated `lib.deno.worker.d.ts`

**Review by bartlomieju — APPROVED** [review]
LGTM, nice work Satya!

File: ext/cache/01_cache.js [hunk 36]
```diff
@@ -0,0 +1,312 @@
...
+                  }
+                } catch (err) {
+                  // There was an error while reading a chunk of the body, so we
+                  // error.
+                  controller.error(err);
+                  controller.close();
+                  core.close(responseBodyRid);
+                }
+              },
+            });
```
Comment (lucacasonato) [inline ext/cache/01_cache.js:277]: This can be replaced with:
```suggestion
            body = readableStreamFromRid(responseBodyRid);
```

**Review by lucacasonato — APPROVED** [review]
LGTM! Great job @satyarohith. 🎉

File: ext/cache/01_cache.js (hunk 36 above)
Comment (satyarohith) [reply ext/cache/01_cache.js:277]: Do you mean to move the code into a function? I couldn't find readableStreamFromRid() defined anywhere.
